rclcpp/test/rclcpp/test_executor.cpp

```diff
@@ -510,24 +511,61 @@ TEST_F(TestExecutor, is_spinning) {
   ASSERT_TRUE(timer_called);
 }
 
-TEST_F(TestExecutor, release_ownership_entity_after_spinning_cancel) {
+class TestAllThreadedExecutor
+  : public ::testing::Test, public ::testing::WithParamInterface<std::string>
+{
+public:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  void TearDown() override
+  {
+    rclcpp::shutdown();
+  }
+};
+
+TEST_P(TestAllThreadedExecutor, release_ownership_entity_after_spinning_cancel) {
   using namespace std::chrono_literals;
+  const std::string single_threaded_executor(
+    typeid(rclcpp::executors::SingleThreadedExecutor).name());
+  const std::string multi_threaded_executor(
+    typeid(rclcpp::executors::MultiThreadedExecutor).name());
 
   // Create an Executor
-  rclcpp::executors::SingleThreadedExecutor executor;
+  auto type_name = GetParam();
+  std::shared_ptr<rclcpp::Executor> executor;
+  if (single_threaded_executor == type_name) {
+    executor = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+  } else if (multi_threaded_executor == type_name) {
+    executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
+  } else {
+    FAIL() << "Unsupported Executor Type !";
+  }
 
-  auto future = std::async(std::launch::async, [&executor] {executor.spin();});
+  auto future = std::async(std::launch::async, [&executor] {executor->spin();});
 
   auto node = std::make_shared<rclcpp::Node>("test_node");
   auto callback = [](
     const test_msgs::srv::Empty::Request::SharedPtr, test_msgs::srv::Empty::Response::SharedPtr) {
     };
   auto server = node->create_service<test_msgs::srv::Empty>("test_service", callback);
-  executor.add_node(node);
+  while (!executor->is_spinning()) {
+    std::this_thread::sleep_for(50ms);
+  }
+  executor->add_node(node);
   std::this_thread::sleep_for(50ms);
-  executor.cancel();
+  executor->cancel();
   std::future_status future_status = future.wait_for(1s);
   EXPECT_EQ(future_status, std::future_status::ready);
 
   EXPECT_EQ(server.use_count(), 1);
 }
+
+INSTANTIATE_TEST_SUITE_P(
+  TestAllThreadedExecutorWithParam,
+  TestAllThreadedExecutor,
+  ::testing::Values(
+    std::string(typeid(rclcpp::executors::SingleThreadedExecutor).name()),
+    std::string(typeid(rclcpp::executors::MultiThreadedExecutor).name())));
```

Comment:
I cannot write the following code because both SingleThreadedExecutor and MultiThreadedExecutor depend on context during construction. If the user runs `rclcpp/test/rclcpp/test_executor --gtest_list`, it will cause an error. This is because when --gtest_list is executed, SetUp() is not called, which will result in an error during the construction of SingleThreadedExecutor and MultiThreadedExecutor.
```
INSTANTIATE_TEST_SUITE_P(
  TestAllThreadedExecutorWithParam,
  TestAllThreadedExecutor,
  ::testing::Values(
    std::make_shared<rclcpp::executors::SingleThreadedExecutor>(),
    std::make_shared<rclcpp::executors::MultiThreadedExecutor>()));
```

Reply:
You should use a type parametrized test.
The executors are still created inside the test, not before.
See as an example https://github.com/ros2/rclcpp/blob/815de21632e27a94c33c89df25e5d89e1ad26d6b/rclcpp/test/rclcpp/executors/test_executors.cpp#L89

Moreover we should run the test also with events executor and static single threaded executor.

Reply:
Got it. Test code should be moved to test_executors.cpp.